### PR TITLE
Add type Alpine

### PR DESCRIPTION
**Pull request by rehhouari**
The type may be helpful when building plugins in Typescript & just using `window.Alpine` in general

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -24,6 +24,29 @@ export abstract class AlpineComponent {
...
+	onBeforeComponentInitializeds: (callback: function) => void;
+	listenForNewUninitializedComponentsAtRunTime: () => undefined;
+	discoverUninitializedComponents: (
+		callback: (rootEl: HTMLElement) => void,
+		el?: HTMLElement = null
+	) => void;
+	discoverComponents: (callback: (rootEl: HTMLElement) => void) => void;
+	start: () => void;
+	addMagicProperty(name: string, callback: ($el: HTMLElement) => void);
+	clone: (component: any, newEl: HTMLElement) => void;
```
Comment (stancl) [inline src/index.ts]: Shouldn't the `component` argument use `AlpineComponent`?

File: src/index.ts [hunk 2]
```diff
@@ -24,6 +24,29 @@ export abstract class AlpineComponent {
...
 
+export declare type Alpine = {
+	version: string;
+	pauseMutationObserver: boolean;
+	magicProperties: { [name: string]: function };
+	ignoreFocusedForValueBinding: boolean;
+	pauseMutationObserver: boolean;
+	onComponentInitializeds: Array<any>;
+	onBeforeComponentInitializeds: Array<any>;
+	onComponentInitializeds: (callback: function) => void;
```
Comment (stancl) [inline src/index.ts]: probably replace all the functions here with `CallableFunction`

File: src/index.ts (hunk 1 above)
Comment (rehhouari) [reply src/index.ts]: Oh yes true!

File: src/index.ts (hunk 2 above)
Comment (rehhouari) [reply src/index.ts]: I wasn't aware of that type, done

File: src/index.ts [hunk 3]
```diff
@@ -24,6 +24,27 @@ export abstract class AlpineComponent {
...
 
+declare type Alpine = {
+	version: string;
+	pauseMutationObserver: boolean;
+	magicProperties: { [name: string]: CallableFunction };
+	ignoreFocusedForValueBinding: boolean;
+	pauseMutationObserver: boolean;
+	onComponentInitializeds: Array<any>;
+	onBeforeComponentInitializeds: Array<any>;
+	onComponentInitializeds: (callback: CallableFunction) => void;
```
Comment (stancl) [inline src/index.ts]: I think the function shouldn't be plural, i.e. `onComponentInitialized`

**Comment by stancl** [comment]
Have you tested this in a project? Does it work well?

It looks correct to my eyes, but want to be sure that nothing is missed.

Thanks!

**Comment by rehhouari** [comment]
> Have you tested this in a project? Does it work well?
> 
> It looks correct to my eyes, but want to be sure that nothing is missed.
> 
> Thanks!

Will test it out on an actual project real quick

File: src/index.ts (hunk 3 above)
Comment (rehhouari) [reply src/index.ts]: yes right sorry!

**Comment by rehhouari** [comment]
Tried it out, and it's working now. will be using it for a couple things

**Comment by stancl** [comment]
Seems like my changes were reverted? I added a `[key: string]: any` catch-all

**Comment by stancl** [comment]
Also, should it be a type or an interface?

File: src/index.ts [hunk 4]
```diff
@@ -24,6 +24,26 @@ export abstract class AlpineComponent {
...
 
+export declare type Alpine = {
+	version: string;
+	pauseMutationObserver: boolean;
+	magicProperties: { [name: string]: CallableFunction };
+	ignoreFocusedForValueBinding: boolean;
+	onComponentInitializeds: Array<CallableFunction>;
+	onBeforeComponentInitializeds: Array<CallableFunction>;
+	onComponentInitialized: (callback: CallableFunction) => void;
+	onBeforeComponentInitialized: (callback: CallableFunction) => void;
```
Comment (stancl) [inline src/index.ts]: I think all of these callbacks could be typed.

Magic properties accept the component el. Initialize hooks accept the component instance (probably) and some of them can also return values iirc. Best to check the Alpine codebase for that

**Comment by rehhouari** [comment]
> Seems like my changes were reverted? I added a `[key: string]: any` catch-all

Oh sorry. I didn't delete it, I'm not sure what exactly happened. Didn't get a change on pull request before.
Did you add it on `Alpine` or `AlpineComponent`? 
Edit: yes my bad, I saw it on AlpineComponent only and I just pasted the entire type over your changes without noticing

> Also, should it be a type or an interface?
Interface is extendable and I think you use it for the component function. so it probably is a better choice?

File: src/index.ts (hunk 4 above)
Comment (rehhouari) [reply src/index.ts]: Oh I see will look into it

**Comment by stancl** [comment]
Looks good now, thanks!

**Pull request merged**